### PR TITLE
Adding automated thread controls for fallback triton servers

### PR DESCRIPTION
**Pull request by wpmccormack**
#### PR description:

For SONIC, the number of instances of a model in a fallback server and the number of threads that fallback server can occupy on local CPU resources can be controlled via the model's triton config file. We set the number of instances to the number of threads specified for the job, so this implementation requires a bit of text editing before the fallback servers can run. The model files live in cvmfs, and the config files cannot be edited there, so they are pulled locally. The precise edits depend on the model's backend; luckily the backend information can be extracted from the config files, so no additional dictionary containing this information is needed.

This PR was tested and able to duplicate results were model config files were adjusted manually.

This PR is also a re-do of https://github.com/cms-sw/cmssw/pull/41228.  That PR was aiming at the wrong target branch.

#### PR validation:

This has been tested locally to verify that: 1) the config files get edited correctly, 2) that the fallback servers start properly, and 3) that a miniAOD processing job can run completely.  It also reproduces some rigorous timings tests that were previously performed with manual config file editing.

### CODE REVIEW

**Comment by wpmccormack** [comment]
@kpedro88 and @makortel , this is a new PR based on https://github.com/cms-sw/cmssw/pull/41228.  Commits have been squashed here, and I rebased to 13_2_0_pre1 just to make things easier

File: HeterogeneousCore/SonicTriton/scripts/cmsTriton [hunk 1]
```diff
@@ -13,12 +13,14 @@ MODELS=()
...
 TMPDIR=""
+INSTANCES=0
+LOCALMODELREPO=""
 DRYRUN=""
 PARENTPID=""
 BASEPORT=8000
 AUTOPORT=""
 NPORTS=3
-IMAGE=fastml/triton-torchgeo:22.07-py3-geometric
+IMAGE=fastml/triton-torchgeo:21.06-py3-geometric
```
Comment (kpedro88) [inline HeterogeneousCore/SonicTriton/scripts/cmsTriton:23]: please fix the rebase, this is a regression.

File: HeterogeneousCore/SonicTriton/scripts/cmsTriton [hunk 2]
```diff
@@ -38,17 +40,18 @@ usage() {
 	$ECHO "-c          \t don't cleanup temporary dir (for debugging)"
 	$ECHO "-C [dir]    \t directory containing Nvidia compatibility drivers (checks CMSSW_BASE by default if available)"
 	$ECHO "-D          \t dry run: print container commands rather than executing them"
-	$ECHO "-d          \t use Docker instead of Apptainer"
+	$ECHO "-d          \t use Docker instead of Singularity"
```
Comment (kpedro88) [inline HeterogeneousCore/SonicTriton/scripts/cmsTriton:43]: please fix the rebase, every change from Apptainer back to Singularity is a regression.

File: HeterogeneousCore/SonicTriton/scripts/cmsTriton (hunk 1 above)
Comment (wpmccormack) [reply HeterogeneousCore/SonicTriton/scripts/cmsTriton]: Fixed

File: HeterogeneousCore/SonicTriton/scripts/cmsTriton (hunk 2 above)
Comment (wpmccormack) [reply HeterogeneousCore/SonicTriton/scripts/cmsTriton]: Yes, sorry I copied that in from an older repository.  Went ahead and edited the new version by hand now

**Comment by kpedro88** [comment]
@wpmccormack it looks like you removed the executable bit from `cmsTriton` (GitHub says `100755 → 100644`), so the unit tests fell back to the base release version of the script (which is no longer compatible with `TritonService`).

**Comment by wpmccormack** [comment]
Yikes, ok I fixed the permissions thing.  The permissions match what they used to be

**Comment by kpedro88** [comment]
please test

**Comment by makortel** [comment]
Workflow 12434.0 shows an occurrence of #41200

**Comment by makortel** [comment]
+heterogeneous

**Comment by kpedro88** [comment]
@wpmccormack here's the documentation: https://github.com/triton-inference-server/server/blob/main/docs/user_guide/model_configuration.md#ensemble-model-instance-groups
> Ensemble Model Instance Groups
> 
> [Ensemble models](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/architecture.md#ensemble-models) are an abstraction Triton uses to execute a user-defined pipeline of models. Since there is no physical instance associated with an ensemble model, the instance_group field can not be specified for it.
> 
> However, each composing model that makes up an ensemble can specify instance_group in its config file and individually support parallel execution as described above when the ensemble receives multiple requests.

and here's the ensemble model config: https://github.com/cms-data/RecoEgamma-EgammaPhotonProducers/blob/4f3629e77af2cafa0f4a8937f859b45e37ca0cb6/models/photonObjectEnsemble/config.pbtxt

I think we need to do two things:
1. Detect when `platform: "ensemble"` (simple)
2. In that case, instead of putting the instance group in the ensemble model config, find each `model_name` in the ensemble and put instance groups in those models' configs (a little more complicated)

**Comment by wpmccormack** [comment]
Ahhh ok great, thanks for finding that.  I'll work on adjusting the implementation and run a few tests

**Pull request merged**